✨Source Okta: migrate to low-code

pabloescoder opened the pull request:
Migrating Source Okta to `YamlDeclarativeSource (Low-code)` Airbyte CDK
Resolves https://github.com/airbytehq/airbyte-internal-issues/issues/6757

No breaking changes, I've tried to replicate everything that was there in the previous python based source. Since the source supported 2 methods of authentication, I have used a SelectiveAuthenticator. Both authentication methods needed the Authorization: SSWS <Token> header, which is not supported by the default Oauth2 and Bearer Authenticators (they add the Authorization: Bearer <Token> header) so I made custom components for the same.

pabloescoder commented on airbyte-integrations/connectors/source-okta/acceptance-test-config.yml:
```diff
@@ -3,7 +3,7 @@ test_strictness_level: high
   spec:
     tests:
-      - spec_path: "source_okta/spec.json"
+      - spec_path: "source_okta/spec.yaml"
```
It doesn't, because when we migrate it to lowcode, we change the json file to yaml, and spec gets inserted in the manifest.yaml file. 
Should I change it to manifest.yaml? Or should I duplicate the spec in a new file spec.yaml?

pabloescoder replied:
Should I just remove the spec_path field then?

pabloescoder replied:
Removed the block as the file didn't exist.

pabloescoder replied:
This is resolved. I merged the latest changes from master and it now shows the full expected and current schema.

pabloescoder commented on airbyte-integrations/connectors/source-okta/main.py:
```diff
@@ -2,7 +2,13 @@
 
 if __name__ == "__main__":
-    run()
+    source = SourceOkta()
```
We can use the run method, this change happened automatically when I ran the generator for low code, should I modify it to use the run method?

natikgadzhi replied:
For consistency with other sources, yes.

pabloescoder replied:
Okay, I'll use the run method

tolik0 commented on airbyte-integrations/connectors/source-okta/unit_tests/test_streams.py:
```diff
@@ -1,399 +0,0 @@
-
-
-@pytest.fixture
-def patch_base_class(mocker):
```
Please return the unit tests. They should be aligned with the new approach instead of being deleted.

pabloescoder replied:
Added unit tests back.

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/config_migration.py:
```diff
@@ -0,0 +1,82 @@
+logger = logging.getLogger("airbyte")
+
+
+class OktaConfigMigration:
```
Please add unit tests for this migration

pabloescoder replied:
Added unit tests for config migration

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py:
```diff
@@ -0,0 +1,65 @@
+
+
+@dataclass
+class CustomOauth2Authenticator(DeclarativeOauth2Authenticator):
```
Please add the unit tests for the custom components

pabloescoder replied:
Added tests for custom components

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:
```diff
@@ -0,0 +1,1882 @@
+version: 0.65.0
```
Please update the version to the latest

pabloescoder replied:
Done.

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:
```diff
@@ -0,0 +1,1882 @@
+    $ref: "#/definitions/groups_stream"
+
+  - type: DeclarativeStream
+    name: group_members
```
Why this stream is no longer incremental?

pabloescoder replied:
group_members stream was not incremental in the python cdk implementation either. 
[Here's a link to the group_members stream class in the Python CDK implementation](https://github.com/airbytehq/airbyte/pull/36509/files#diff-a18919b9baef2ea667a3623bf4b35de40c9f6aa5f2bf8816bb54e22470aba01fL122)

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:
```diff
@@ -0,0 +1,1882 @@
+    $parameters:
+      name: "group_members"
+    primary_key:
+      - type
```
Why have you changed the primary key? This is a breaking change

pabloescoder replied:
Fixed this, the primary key is back to the one used in the Python implementation

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:
```diff
@@ -0,0 +1,1882 @@
+    $parameters:
+      name: "group_role_assignments"
+    primary_key:
+      - id
```
Why have you changed the primary key?

pabloescoder replied:
My apologies. Fixed it.

tolik0 commented on airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:
```diff
@@ -0,0 +1,1882 @@
+    $parameters:
+      name: "user_role_assignments"
+    primary_key:
+      - id
```
Why have you changed the primary key?

pabloescoder replied:
Fixed this.

pabloescoder commented:
@tolik0 
I've added the unit tests back. Added new unit tests for config migration, fixed the issue with the wrong primary keys, updated airbyte-cdk and expected records. 
This PR is ready for review.

Also, could you please run airbyte-ci format locally and push those changes here? 
I've ran the `airbyte-ci format fix all` command 4-5 times and even though it runs successfully locally, it is failing here in the pipelines.

lazebnyi commented:
/format-fix

pabloescoder commented:
The check for formatting errors pipeline is failing only on the unit_tests folder (which is excluded from the airbyte-ci path for formatting - [see here](https://github.com/airbytehq/airbyte/blob/master/pyproject.toml#L31C5-L31C17))
All the other files are properly formatted so we can ignore that.
